sydent/util/hash.py

```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+import unpaddedbase64
+
+
+def sha256_and_url_safe_base64(input_text):
+    """SHA256 hash an input string, encode the digest as url-safe base64, and
+    return
+
+    :param input_text: string to hash
+    :type input_text: str
+
+    :returns a sha256 hashed and url-safe base64 encoded digest
+    :rtype: str
+    """
+    digest = hashlib.sha256(input_text.encode()).digest()
+    return unpaddedbase64.encode_base64(digest, urlsafe=True)
+
+def parse_space_separated_str(input_str):
+    """Parses a string containing values seperated by a space. Joins the
+    leading chunks if there are more than two.
```

Comment:
This feels quite a weird thing to want tbh. Do we have any addresses that have a space? Do we instead want to assert the addresses *don't* have an extra space, to try and catch errors?

Reply:
Spaces are allowed in emails because of course they are: https://en.wikipedia.org/wiki/Email_address#Local-part

Reply:
Ugh, that's likely to bite us if we are using space as a separator. I would be *very* tempted to still ban spaces on the assumption I doubt anyone actually uses spaces.

Reply:
Ugh, that's likely to bite us if we are using space as a separator. I would be *very* tempted to still ban spaces on the assumption I doubt anyone actually uses spaces.

Reply:
Yeah, practically we can prooooobably ignore it.

It was also in case we ever come across a service that uses spaces in their addresses, and we add that as a valid third-party provider. But we can cross that bridge when we come to it. No database migration should be needed.

Reply:
Spaces in addresses are now disallowed.